rfd/0210-windows-ad-hsm.md

```diff
@@ -0,0 +1,198 @@
+
+The assumption that there is a single CA is invalid when Teleport is configured
+to use HSMs, as each auth server will have its own certificate and HSM-backed
+private key. Windows requires that the CRL pointed to by the certificate is
+signed by the same CA as the certificate itself. As a result, Teleport logins
+only work 1 out of N times (where N is the number of auth servers in the
+cluster) when configured for HSMs, because the auth server that issued the
+cert for the session or may not be the same as the auth server that generated
+the CRL.
+
+## Required Approvers
+
+* Engineering: @nklaassen && (@gabrielcorado || @greedy52)
+* Security: (@rjones || @klizhentas)
+
+## Details
+
+Windows supports a variety of different CRL distribution points:
+
+- local paths: `file://...`
+- remote file share paths: /`\\server\share\...`
+- LDAP objects: `ldap://CN=Teleport,CN=CDP,...`
+- HTTP/HTTPS endpoints: `https://...`
+
+Teleport has always published its CRLs to LDAP, as it's guaranteed to be
+available in the environments where Teleport runs and Teleport already has an
+LDAP requirement. While it would be possible to make the Teleport proxy server
+host CRLs over HTTPS, we aim to avoid such a change as we can't be sure that all
+of the Windows components have network access to the Teleport proxy.
+
+> [!NOTE]
+> Teleport has its own locking system and does not use CRLs for revocation.
+> Teleport only implements CRLs to satisfy Windows requirements, and its CRLs
+> are always empty and always valid for 1 year from time of issuance.
+
+In order to support HSM-enabled clusters with multiple active signers, we need
+to consider two aspects:
+
+1. Certificate generation: how does Teleport generate the CRL distribution point
+2. CRL publishing: how does the CRL actually get to the distribution point referenced in the cert
+
+### Certificate Generation
+
+The certificate generation step is largely the same for Windows desktop access
+and MS SQL database access. In both cases, Teleport generates an LDAP URL based
+on the configured Active Directory domain. It takes the form:
+
+```
+ldap://CN=CLUSTER,CN=SIGNER,CN=CDP,CN=Public Key Services,CN=Configfuration,DC=example,DC=com
+```
+
+Where:
+- `CLUSTER` is the name of the Teleport cluster
+- `SIGNER` is `Teleport` (for Windows desktop certs) or `TeleportDB` for database certs
+- `example.com` is the AD domain (`DC=example,DC=com` in LDAP-speak)
+
+### CRL Publishing
+
+CRL publishing works differently for each Teleport feature.
+
+#### MS SQL database access
+
+For database access, our setup instructions require that users generate the CRL
+with `tctl auth crl`, copy it over to the Windows environment, and publish it to
+LDAP using `certutil -dspublish`.
+
+It is worth noting that `tctl auth crl` generates empty CRLs that are valid for _1 year_,
+so database access will require manual intervention in case:
+
+- the CRL expires after 1 year
+- a CA rotation is performed
```

Comment:
We could improve the UX here quite a lot by using the automated publishing workflow. We should probably keep the `tctl auth crl` as a fallback.

We will likely need an active LDAP connectivity anyway to lookup SID for username in order to fix https://github.com/gravitational/teleport/issues/53502.

Reply:
Going to keep that out of scope for now, since getting HMS to work is priority over other UX improvements. Feel free to file a new issue and link to this RFD for book-keeping.

Reply:
Oh sure, this was just a side remark, no need to change the RFD.